cudax/include/cuda/experimental/__container/async_buffer.cuh

```diff
@@ -0,0 +1,862 @@
+  template <class _Range>
+  static constexpr bool __compatible_range = _CUDA_VRANGES::__container_compatible_range<_Range, _Tp>;
+
+  //! @brief Helper to check whether a different async_buffer still satisfies all properties of this one
+  template <class... _OtherProperties>
+  static constexpr bool __properties_match =
+    !_CCCL_TRAIT(_CUDA_VSTD::is_same,
+                 _CUDA_VSTD::__make_type_set<_Properties...>,
+                 _CUDA_VSTD::__make_type_set<_OtherProperties...>)
+    && _CUDA_VSTD::__type_set_contains_v<_CUDA_VSTD::__make_type_set<_OtherProperties...>, _Properties...>;
+
+  //! @brief Helper to determine what cudaMemcpyKind we need to copy data from another async_buffer with different
+  template <class... _OtherProperties>
+  static constexpr cudaMemcpyKind __transfer_kind =
+    __select_execution_space<_OtherProperties...> == _ExecutionSpace::__host
+      ? (__is_host_only ? cudaMemcpyHostToHost : cudaMemcpyHostToDevice)
+      : (__is_host_only ? cudaMemcpyDeviceToHost : cudaMemcpyDeviceToDevice);
+
+  //! @brief Helper to return an async_resource_ref to the currently used resource. Used to grow the async_buffer
+  __resource_ref_t __borrow_resource() const noexcept
+  {
+    return const_cast<__resource_t&>(__buf_.get_memory_resource());
+  }
+
+  //! @brief Replaces the content of the async_buffer with the sequence `[__first, __last)`
+  //! @param __first Iterator to the first element of the input sequence.
+  //! @param __last Iterator after the last element of the input sequence.
+  template <class _Iter, cudaMemcpyKind __kind = __detect_transfer_kind<__is_host_only, _Iter>>
+  _CCCL_HIDE_FROM_ABI void __assign_impl(const size_type __count, _Iter __first, _Iter __last)
+  {
+    if (size() < __count)
+    {
+      (void) __buf_.__replace_allocation(__count);
+    }
+
+    this->__copy_cross<_Iter, __kind>(__first, __last, __unwrapped_begin(), __count);
+    __size_ = __count;
+  }
+
+  //! @brief Copies \p __count elements from `[__first, __last)` to \p __dest, where \p __first and \p __dest reside in
+  //! the different memory spaces
+  //! @param __first Pointer to the start of the input segment.
+  //! @param __last Pointer to the end of the input segment.
+  //! @param __dest Pointer to the start of the output segment.
+  //! @param __count The number of elements to be copied.
+  //! @note This function is inherently asynchronous. We need to ensure that the memory pointed to by \p __first and
+  //! \p __last lives long enough
+  template <class _Iter, cudaMemcpyKind __kind = __is_host_only ? cudaMemcpyHostToHost : cudaMemcpyHostToDevice>
+  _CCCL_HIDE_FROM_ABI void __copy_cross(_Iter __first, _Iter __last, pointer __dest, size_type __count)
+  {
+    if constexpr (__kind == cudaMemcpyHostToHost)
+    {
+      ::cuda::experimental::host_launch(__buf_.get_stream(), _CUDA_VSTD::copy<_Iter, pointer>, __first, __last, __dest);
+      __buf_.get_stream().wait();
```

Comment:
Important: The `wait()` should not be necessary, since the stream-ordered callback does not depend on anything here. Was a bug observed because of the missing wait? If yes, it could hint at a design problem with the API allowing a bug to happen easily. I think we need to understand this.